BUG: Fix file descriptor leak

roberthdevries opened the pull request:
- [x] closes #31488 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

roberthdevries committed:
Fix file descriptor leak

roberthdevries committed:
Add whatsnew entry

gfyoung commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -314,3 +316,15 @@ def test_malformed_skipfooter(python_parser_only):
         parser.read_csv(StringIO(data), header=1, comment="#", skipfooter=1)
+
+
+def test_file_descriptor_leak(python_parser_only):
```
* Can't we also test this against the C parser?

roberthdevries replied:
No problem, done.

gfyoung commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -314,3 +316,15 @@ def test_malformed_skipfooter(python_parser_only):
+    # GH 31488
+    parser = python_parser_only
+    with open("empty.csv", "w"):
+        pass
```
Can we try using our `ensure_clean` utility instead?

gfyoung commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -314,3 +316,15 @@ def test_malformed_skipfooter(python_parser_only):
+def test_file_descriptor_leak(python_parser_only):
+    # GH 31488
+    parser = python_parser_only
+    with open("empty.csv", "w"):
```
Can we try using our ensure_clean utility instead?

roberthdevries replied:
Fixed

gfyoung commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -314,3 +316,15 @@ def test_malformed_skipfooter(python_parser_only):
+    proc = psutil.Process()
+    with pytest.raises(EmptyDataError):
+        parser.read_csv("empty.csv")
+    assert not proc.open_files()
```
I would check that the list of open files does not contain our file, not that it's empty (more vulnerable to failure if the testing environment happen to have other files open).

roberthdevries replied:
I have now changed the test to a before/after check that triggers when the results differ.

gfyoung commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -314,3 +316,15 @@ def test_malformed_skipfooter(python_parser_only):
+        pass
+
+    proc = psutil.Process()
+    with pytest.raises(EmptyDataError):
```
Check the error message as well.

roberthdevries replied:
Done

roberthdevries committed:
Use ensure_clean

roberthdevries committed:
Add the expected exception message

roberthdevries committed:
isort fixes

roberthdevries committed:
Made the test non-python parser specific

roberthdevries committed:
import psutil fails on several travis runs due to missing psutil

roberthdevries committed:
Fix import order

roberthdevries committed:
Fix for black formatter

WillAyd commented on pandas/tests/io/parser/test_common.py:
```diff
@@ -2079,3 +2080,17 @@ def test_integer_precision(all_parsers):
+
+
+@td.skip_if_no("psutil")
+def test_file_descriptor_leak(all_parsers):
```
We already have some machinery to wrap the function and check for a leak here:

https://github.com/pandas-dev/pandas/blob/25e2f8687f0a7033ddbbd6aa1b833d0586f99bad/pandas/util/_test_decorators.py#L236

Can you use this instead?

roberthdevries replied:
I tried it, but it doesn't trigger on this specific issue. Apparently the file handle gets destroyed by some other means before the check can trigger.
But I can double check.

roberthdevries replied:
I can choose between using the `tm.ensure_clean` context manager which will close the file as well and therefore prevent the file descriptor leak, or use the `check_file_leaks` context manager, but not both.
Please express a preference of one above the other.

WillAyd replied:
I think should use the `check_file_leaks` context manager

roberthdevries replied:
OK, so that means I have to ditch the `tm.ensure_clean` call?

WillAyd replied:
I'm not clear why they can't both be used together - what error were you getting?

roberthdevries replied:
Well, the `tm.ensure_clean` basically resolves the file descriptor leaked by `read_csv()`, so that the enclosing `check_file_leaks` does not trigger.
I always make sure that the test fails when I undo my change. And with both in place, the test succeeds even when the bug is present. Therefore the test becomes useless when both constructs are used in combination.

WillAyd replied:
How are you setting that up? I see this as raising on master:

```python
with tm.ensure_clean() as path:
    def shouldnt_leak():
        try:
            pd.read_csv(path)
        except EmptyDataError:
            pass
        
    check_file_leaks(shouldnt_leak)()
```

Though admittedly intermittently

roberthdevries replied:
That could work, because now the file leak checking is inside the `tm.ensure_clean` context manager.
I used it as a decorator of the test in this fashion:
```python
@td.check_file_leaks
def test_bla():
    with tm.ensure_clean() as path:
        try:
            pd.read_csv(path)
        except EmptyDataError:
            pass
```

roberthdevries replied:
This seems to work for me, so it is now changed to use td.check_file_leaks in the right manner.

WillAyd reviewed (changes requested):
Thanks! Looks good just some standardization on the test that can be improved

jreback commented on pandas/io/parsers.py:
```diff
@@ -2273,11 +2273,15 @@ def __init__(self, f, **kwds):
+            ) = self._infer_columns()
+        except EmptyDataError:
+            self.close()
+            raise
```
this should be

```
try: 
......
except EmptyDataError as err:
      self.close()
      raise from err
```

jreback replied:
also I think ok to catch TypeError, ValueError, EmptyDataError here (all things raised by infer_columns)

roberthdevries replied:
The raise from clause only works in the case when you change the exception type, not in this case, so if you do `raise MyNewException from err`
`EmptyDataError` is derived from `ValueError` so I changed it to catch `ValueError` and `TypeError`

jreback commented on pandas/tests/io/parser/test_common.py:
```diff
@@ -2079,3 +2080,17 @@ def test_integer_precision(all_parsers):
+    # GH 31488
+    import psutil
+
+    proc = psutil.Process()
```
can you add a comment here very similar to what you posted above (e.g. why using psutil open files to check)

roberthdevries replied:
added

roberthdevries committed:
Add TypeError and ValueError to the list of caught exceptions

Also add a comment why combining tm.ensure_clean and td.check_file_leaks
don't play along nicely together

roberthdevries committed:
Use td.check_file_leaks() in a way that works in combination with tm.ensure_clean

jreback commented:
thanks @roberthdevries
